File: examples/example-kubernetes-client-openapi-operator-python/src/test/python/micronaut/operator/ConfigMapReconcilerTest.py
from java.util import Optional
from micronaut.kubernetes.client.openapi.model import V1ConfigMap, V1ObjectMeta
from micronaut.kubernetes.client.openapi.operator.controller.reconciler import Request
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from micronaut.operator.ConfigMapReconciler import ConfigMapReconciler


class RecordingCoreV1Api:
    """Records the config maps the reconciler replaces instead of calling the Kubernetes API."""

    def __init__(self):
        self.replaced = []

    def replaceNamespacedConfigMap(self, name, namespace, body, pretty, dry_run, field_manager, field_validation):
        self.replaced.append((name, namespace, body))
        return body


class SingleEntryLister:
    """Stands in for the OperatorResourceLister backed by the informer cache."""

    def __init__(self, config_map: V1ConfigMap):
        self.config_map = config_map

    def get(self, request: Request) -> Optional[V1ConfigMap]:
        metadata = self.config_map.getMetadata()
        if metadata.getName() == request.name() and metadata.getNamespace() == request.namespace():
            return Optional.of(self.config_map)
        return Optional.empty()


def config_map(name: str, namespace: str) -> V1ConfigMap:
    metadata = V1ObjectMeta()
    metadata.setName(name)
    metadata.setNamespace(namespace)
    cm = V1ConfigMap()
    cm.setMetadata(metadata)
    return cm


@MicronautTest
class ConfigMapReconcilerTest:

    @Test
    def test_reconcile_annotates_the_config_map(self):
        cm = config_map("my-config", "default")
        lister = SingleEntryLister(cm)
        api = RecordingCoreV1Api()
        reconciler = ConfigMapReconciler(api)

        result = reconciler.reconcile(Request("my-config", "default"), lister)

        assert not result.requeue()
        assert len(api.replaced) == 1
        assert cm.getMetadata().getAnnotations().get("io.micronaut.operator") == "processed"

        # the annotation is already present, so the second reconciliation does not update the resource
        result = reconciler.reconcile(Request("my-config", "default"), lister)
        assert not result.requeue()
        assert len(api.replaced) == 1

    @Test
    def test_reconcile_of_a_deleted_config_map(self):
        lister = SingleEntryLister(config_map("other", "default"))
        api = RecordingCoreV1Api()

        result = ConfigMapReconciler(api).reconcile(Request("my-config", "default"), lister)

        assert not result.requeue()
        assert len(api.replaced) == 0

File: examples/example-kubernetes-client-openapi-operator-kotlin/src/test/kotlin/micronaut/operator/ConfigMapReconcilerTest.kt
package micronaut.operator

import io.micronaut.kubernetes.client.openapi.api.CoreV1Api
import io.micronaut.kubernetes.client.openapi.informer.SharedIndexInformer
import io.micronaut.kubernetes.client.openapi.informer.SharedIndexInformerFactory
import io.micronaut.kubernetes.client.openapi.informer.cache.Indexer
import io.micronaut.kubernetes.client.openapi.model.V1ConfigMap
import io.micronaut.kubernetes.client.openapi.model.V1ObjectMeta
import io.micronaut.kubernetes.client.openapi.operator.OperatorResourceLister
import io.micronaut.kubernetes.client.openapi.operator.controller.reconciler.Request
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.isNull
import org.mockito.Mockito.mock
import org.mockito.Mockito.never
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.mockito.Mockito.`when`

class ConfigMapReconcilerTest {

    private fun configMap(name: String) = V1ConfigMap().metadata(V1ObjectMeta().name(name).namespace("default"))

    private fun lister(vararg configMaps: V1ConfigMap): OperatorResourceLister<V1ConfigMap> {
        val indexer = mock(Indexer::class.java) as Indexer<V1ConfigMap>
        for (configMap in configMaps) {
            `when`(indexer.getByKey("default/" + configMap.metadata!!.name)).thenReturn(configMap)
        }
        val informer = mock(SharedIndexInformer::class.java) as SharedIndexInformer<V1ConfigMap>
        `when`(informer.indexer).thenReturn(indexer)
        val factory = mock(SharedIndexInformerFactory::class.java)
        `when`(factory.getExistingSharedIndexInformer(V1ConfigMap::class.java, "default")).thenReturn(informer)
        return OperatorResourceLister(factory, V1ConfigMap::class.java, false)
    }

    @Test
    fun testReconcileAnnotatesTheConfigMap() {
        val configMap = configMap("my-config")
        val lister = lister(configMap)
        val coreV1Api = mock(CoreV1Api::class.java)
        val reconciler = ConfigMapReconciler(coreV1Api)

        var result = reconciler.reconcile(Request("my-config", "default"), lister)

        assertFalse(result.requeue())
        assertEquals("processed", configMap.metadata!!.annotations!!["io.micronaut.operator"])
        verify(coreV1Api, times(1)).replaceNamespacedConfigMap("my-config", "default", configMap, null, null, null, null)

        // the annotation is already present, so the second reconciliation does not update the resource
        result = reconciler.reconcile(Request("my-config", "default"), lister)
        assertFalse(result.requeue())
        verify(coreV1Api, times(1)).replaceNamespacedConfigMap("my-config", "default", configMap, null, null, null, null)
    }

    @Test
    fun testReconcileOfADeletedConfigMap() {
        val coreV1Api = mock(CoreV1Api::class.java)

        val result = ConfigMapReconciler(coreV1Api).reconcile(Request("my-config", "default"), lister(configMap("other")))

        assertFalse(result.requeue())
        verify(coreV1Api, never()).replaceNamespacedConfigMap(any(), any(), any(), isNull(), isNull(), isNull(), isNull())
    }
}
